anonlink/solving.py

```diff
@@ -47,51 +49,85 @@ def greedy_solve(
     assert dset_indices.shape[1] == sorted_indices.shape[0]
     assert rec_indices.shape[1] == sorted_indices.shape[0]
 
-    matches = {}  # type: Dict[Tuple[int, int], Set[Tuple[int,int]]]
-    matchable_pairs = set()
+    # Map (dataset index, record index) to its group.
+    matches = {}  # type: Dict[Tuple[int, int], List[Tuple[int,int]]]
+
+    # Each group is a set of records. We merge two groups every pair of
+    # their records is matchable. A pair is matchable if we have seen
+    # it. Store the number of matchable pairs between two groups. This
+    # is a sparse matrix: the default number of matchable pairs is 0.
+    # Since the groups themselves are not hashable, we use their id as
+    # the key.
+    matchable_pairs = collections.defaultdict(
+        collections.Counter)  # type: DefaultDict[int, Counter[int]]
+
     for dset_is, rec_is in zip(dset_indices.T, rec_indices.T): 
         i0, i1 = zip(dset_is, rec_is)
 
         if i0 in matches and i1 in matches:
             # Both records are assigned to a group.
+            i0_matches = matches[i0]
+            i1_matches = matches[i1]
+            i0_mid = id(i0_matches)
+            i1_mid = id(i1_matches)
+
             # Check if mergeable.
-            matchable_pairs.add(frozenset((i0, i1)))
-            if all(frozenset((j0, j1)) in matchable_pairs
-                   for j0 in matches[i0]
-                   for j1 in matches[i1]):
-                # Yes, mergeable.
-                # Minor cleanup.
-                matchable_pairs.difference(frozenset((j0, j1))
-                                             for j0 in matches[i0]
-                                             for j1 in matches[i1])
+            if (matchable_pairs[i0_mid][i1_mid] + 1
+                == len(i0_matches) * len(i1_matches)):
```

Comment:
Please add a comment clarifying why this condition implies that the two groups can be merged.

Reply:
Done!